stdlib/functools.pyi

```diff
@@ -72,22 +74,48 @@ class _CacheInfo(NamedTuple):
     currsize: int
 
 @final
-class _lru_cache_wrapper(Generic[_T]):
-    __wrapped__: Callable[..., _T]
-    def __call__(self, *args: Hashable, **kwargs: Hashable) -> _T: ...
+class _lru_cache_wrapper_0(Generic[_P, _R]):
+    @property
+    def __wrapped__(self) -> Callable[_P, _R]: ...
+    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _R: ...
+    def cache_info(self) -> _CacheInfo: ...
+    def cache_clear(self) -> None: ...
+    def __copy__(self) -> _lru_cache_wrapper_0[_P, _R]: ...
+    def __deepcopy__(self, __memo: Any) -> _lru_cache_wrapper_0[_P, _R]: ...
+
+@final
+class _lru_cache_wrapper(Generic[_S, _P, _R]):
+    @property
+    def __wrapped__(self) -> Callable[Concatenate[_S, _P], _R]: ...
+    def __call__(self, arg0: _S, *args: _P.args, **kwargs: _P.kwargs) -> _R: ...
     def cache_info(self) -> _CacheInfo: ...
     def cache_clear(self) -> None: ...
-    def __copy__(self) -> _lru_cache_wrapper[_T]: ...
-    def __deepcopy__(self, __memo: Any) -> _lru_cache_wrapper[_T]: ...
+    def __copy__(self) -> _lru_cache_wrapper[_S, _P, _R]: ...
+    def __deepcopy__(self, __memo: Any) -> _lru_cache_wrapper[_S, _P, _R]: ...
+    def __get__(self, inst: object, owner: type) -> _lru_cache_wrapper_0[_P, _R]: ...
+
+class _LRUCacheDecorator(Protocol):
+    @overload
+    def __call__(self, func: Callable[Concatenate[_S, _P], _R]) -> _lru_cache_wrapper[_S, _P, _R]: ...
+    @overload
+    def __call__(self, func: Callable[_P, _R]) -> _lru_cache_wrapper_0[_P, _R]: ...
 
 if sys.version_info >= (3, 8):
     @overload
-    def lru_cache(maxsize: int | None = ..., typed: bool = ...) -> Callable[[Callable[..., _T]], _lru_cache_wrapper[_T]]: ...
+    def lru_cache(maxsize: int | None = ..., typed: bool = ...) -> _LRUCacheDecorator: ...
+    @overload
+    def lru_cache(maxsize: Callable[Concatenate[_S, _P], _R], typed: bool = ...) -> _lru_cache_wrapper[_S, _P, _R]: ...
     @overload
-    def lru_cache(maxsize: Callable[..., _T], typed: bool = ...) -> _lru_cache_wrapper[_T]: ...
+    def lru_cache(maxsize: Callable[_P, _R], typed: bool = ...) -> _lru_cache_wrapper_0[_P, _R]: ...
```

Comment:
Even if this didn't cause a mypy crash, don't these overloads overlap? How would a type checker ever know to choose the third overload over the second overload?

Reply:
here's some examples which cannot match the first overload but can match the second:

- `def f(**k):`  (no positional argument to concatenate)
- `def f(*, k):` (no positional argument to concatenate)
- `def f():` (no positional argument)

Reply:
Right, because a `ParamSpec` can be bound to an "empty" parameter specification, but `Concatenate[_S, _P]` cannot. Thanks.